orbotservice/src/main/java/org/torproject/android/service/circumvention/AutoConf.kt

```diff
@@ -0,0 +1,160 @@
+package org.torproject.android.service.circumvention
+
+import IPtProxy.IPtProxy
+import android.content.Context
+import java.net.Authenticator
+import java.net.PasswordAuthentication
+
+object AutoConf {
+
+    private const val MEEK_PARAMETERS = "url=https://1723079976.rsc.cdn77.org;front=www.phpmyadmin.net"
+
+    /**
+     * Tries to automatically configure Pluggable Transports, if the MOAT service decides, that in your country one is needed.
+     *
+     * @param country: An ISO 3166-1 Alpha 2 country code to force a specific country.
+     * 	 If not provided, the MOAT service will deduct a country from your IP address (preferred!)
+     * @param cannotConnectWithoutPt: Set to `true`, if you are sure, that a PT configuration *is needed*,
+     * 	 even though the MOAT service says, that in your country none is. In that case, a default configuration will be used.
+     * @return A transport to use and a list of custom bridges, if any could be evaluated.
+     *
+     * @throws MoatApi.MoatError or maybe other IO exceptions.
+     */
+    suspend fun `do`(
+        context: Context,
+        country: String? = null,
+        cannotConnectWithoutPt: Boolean = false
+    ): Pair<Transport, List<String>>? {
+        var cannotConnectWithoutPt = cannotConnectWithoutPt
+
+        val done = fun(conf: Pair<Transport, List<String>>?): Pair<Transport, List<String>>? {
+            Transport.controller.stop(IPtProxy.MeekLite)
```

Comment:
Here's a sneaky one, in the event that an exception is thrown such as when the Toast with the localized exception message is displayed that I was talking about on the main thread, this `done` function is **not** invoked. Meek is only turned off when things go smoothly, but never on errors. Verified this with breakpoints on this line and the line to start meek below.

Reply:
Uuuh. Yeah, that is also a great catch.
I got myself lulled by Retrofit. But, of course, the Retrofit API calls can throw all sorts of exceptions.
I'll catch them and make sure, the Meek connection is closed. Thank you!